core/src/Kokkos_Core.hpp

```diff
@@ -49,6 +49,22 @@
 #define KOKKOS_IMPL_PUBLIC_INCLUDE_NOTDEFINED_CORE
 #endif
 
+//----------------------------------------------------------------------------
+// In the case windows.h is included before Kokkos_Core.hpp there might be
+// errors due to the potentially defined macros with name "min" and "max" in
+// windows.h. These collide with the use of "min" and "max" in names inside
+// Kokkos. The macros will be redefined at the end of Kokkos_Core.hpp
+#if defined(min)
+#pragma push_macro("min")
+#undef min
+#define KOKKOS_IMPL_PUSH_MACRO_MIN
+#endif
+#if defined(max)
+#pragma push_macro("max")
+#undef max
+#define KOKKOS_IMPL_PUSH_MACRO_MAX
+#endif
```

Comment:
Let's only do that when we could possibly use the windows runtime, i.e, guard it with `#if defined(_WIN32)`.

Reply:
My reasoning was, that the problem is not windows specific but just becomes painfully apparent when including `windows.h`. Now we don't want to guard against any name being used by us ... but is there a good reason to restrict it to windows only?

Reply:
We are dealing with a lot of different compilers and `push_macro` is not standardized (even though it's supported by many compilers). That's why I would rather want to be careful and avoid using it without checking that the compiler actually supports it.

Reply:
makes sense, but it checks if `min` or `max` are defined and only then does the push. If we can not push the defined macros the compiler will fail anyways later. With the given implementation it will at least point to a line that actually explains what is wrong (the macro being defined).

Reply:
I'm initiall inclined to disagree - anything that's defined macros for those common function names needs to be protected against.

On the other hand, if we see such definitions without `defined _WIN32`, then perhaps we should just error out explicitly, because we're on an even weirder platform, or the user is doing something that will screw us up

Reply:
On the flip side, we could just guard all of the definitions and uses inside Kokkos with parentheses, like HPX did, and not mess around with macros, whether they're present or not.

Reply:
if the user wants to define something like `min` or `max` macro why should he not be allowed to use it? My intent was just to protect all our definitions in `Kokkos_Core.hpp` to not suffer from the problem without making a ton of changes by putting `()` everywhere. For future development workflow I think it is nice to not have to remember to put `()`.